File: Lexer/Java/src/main/java/com/code/lexer/Lexer.java
package com.code.lexer;

import com.code.lexer.Token.TokenType;

public class Lexer {
    private String input;
    private int curr;
    private int peek;
    private char ch;

    Lexer(String input) {
        this.input = input;
        this.curr = 0;
        this.peek = 0;
        this.ch = '\0';

        this.readChar();
    }

    public Token nextToken() {
        Token tok = new Token(TokenType.EOF, "\0");

        while (Character.isWhitespace(this.ch))
            this.readChar();

        switch (this.ch) {
            case '\0':
                return tok;
            case '*':
                tok = new Token(TokenType.ASTERISK, "*");
                break;
            case '-':
                tok = new Token(TokenType.DASH, "-");
                break;
            case '+':
                tok = new Token(TokenType.PLUS, "+");
                break;
            case '=':
                tok = new Token(TokenType.EQUAL, "=");
                break;
            case '<':
                tok = new Token(TokenType.LESSTHAN, "<");
                break;
            case '>':
                tok = new Token(TokenType.GREATERTHAN, ">");
                break;
            case '/':
                tok = new Token(TokenType.SLASH, "/");
                break;
            case ';':
                tok = new Token(TokenType.SEMICOLON, ";");
                break;
            case ':':
                tok = new Token(TokenType.COLON, ":");
                break;
            case '(':
                tok = new Token(TokenType.LPAREN, "(");
                break;
            case ')':
                tok = new Token(TokenType.RPAREN, ")");
                break;
            case '{':
                tok = new Token(TokenType.LBRACE, "{");
                break;
            case '}':
                tok = new Token(TokenType.RBRACE, "}");
                break;
            default:
                if (Character.isAlphabetic(this.ch)) {
                    String ident = this.readIdentifier();
                    TokenType ttype = Token.IDENTIFIER_MAP.get(ident);
                    return ttype != null ? new Token(ttype, ident) : new Token(TokenType.IDENTIFIER, ident);
                } else if (Character.isDigit(this.ch)) {
                    return new Token(TokenType.NUMBER, this.readNumber());
                } else {
                    return new Token(TokenType.ILLEGAL, "ILLEGAL");
                }
        }
        this.readChar();
        return tok;
    }

    private String readIdentifier() {
        int pos = this.curr;
        while (Character.isLetter(this.ch) || this.ch == '_')
            this.readChar();
        return this.input.substring(pos, this.curr);
    }

    private String readNumber() {
        int pos = this.curr;
        while (Character.isDigit(this.ch))
            this.readChar();
        return this.input.substring(pos, this.curr);
    }

    private void readChar() {
        this.ch = this.peek >= this.input.length() ? '\0' : this.input.charAt(this.peek);
        this.curr = this.peek++;
    }
}

File: Lexer/Go/lexer/lexer.go
package lexer

import (
	token "golex/token"
	"unicode"
)

type Lexer struct {
	Curr  int
	Peek  int
	Char  byte
	Input string
}

func New(input string) *Lexer {
	lex := Lexer{Input: input}
	lex.readChar()
	return &lex
}

func (l *Lexer) NextToken() token.Token {
	for unicode.IsSpace(rune(l.Char)) {
		l.readChar()
	}
	var tok token.Token

	switch l.Char {
	case '*':
		tok = token.New(token.ASTERISK, "*")
	case '-':
		tok = token.New(token.DASH, "-")
	case '+':
		tok = token.New(token.PLUS, "+")
	case '=':
		tok = token.New(token.EQUAL, "=")
	case '<':
		tok = token.New(token.LESSTHAN, "<")
	case '>':
		tok = token.New(token.GREATERTHAN, ">")
	case '/':
		tok = token.New(token.SLASH, "/")
	case ';':
		tok = token.New(token.SEMICOLON, ";")
	case ':':
		tok = token.New(token.COLON, ":")
	case '(':
		tok = token.New(token.LPAREN, "(")
	case ')':
		tok = token.New(token.RPAREN, ")")
	case '{':
		tok = token.New(token.LBRACE, "{")
	case '}':
		tok = token.New(token.RBRACE, "}")
	case 0:
		tok = token.New(token.EOF, "")
	default:
		if isLetter(l.Char) {
			ident := l.readIdentifier()
			if ttype, ok := token.IdentifierMap[ident]; ok {
				tok = token.New(ttype, ident)
			} else {
				tok = token.New(token.IDENTIFIER, ident)
			}
			return tok
		} else if isNumber(l.Char) {
			return token.New(token.NUMBER, l.readNumber())
		} else {
			tok = token.New(token.ILLEGAL, "ILLEGAL")
		}
	}
	l.readChar()

	return tok
}

func (l *Lexer) readIdentifier() string {
	pos := l.Curr
	for isLetter(l.Char) || l.Char == '_' {
		l.readChar()
	}
	return l.Input[pos:l.Curr]
}

func (l *Lexer) readNumber() string {
	pos := l.Curr
	for isNumber(l.Char) {
		l.readChar()
	}
	return l.Input[pos:l.Curr]
}

func (l *Lexer) readChar() {
	if l.Peek >= len(l.Input) {
		l.Char = 0
	} else {
		l.Char = l.Input[l.Peek]
	}
	l.Curr = l.Peek
	l.Peek += 1
}

func isLetter(ch byte) bool {
	// 'a' <= c <= 'z' OR 'A' <= c <= 'Z'
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z'
}

func isNumber(ch byte) bool {
	return ch >= '0' && ch <= '9'
}
